In [ ]:
install.packages("nycflights13")
install.packages("tidyverse")
# install.packages("dplyr")

library("nycflights13")
library("tidyverse")
# library("dplyr")



Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘textshaping’, ‘conflicted’, ‘jsonlite’, ‘ragg’, ‘stringr’


Warning message in install.packages("tidyverse"):
“installation of package ‘textshaping’ had non-zero exit status”
Warning message in install.packages("tidyverse"):
“installation of package ‘ragg’ had non-zero exit status”
Warning message in install.packages("tidyverse"):
“installation of package ‘tidyverse’ had non-zero exit status”
Warning message in system("timedatectl", intern = TRUE):
“running command 'timedatectl' had status 1”
── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──

✔ ggplot2 3.4.1     ✔ purrr   1.0.1
✔ tibble  3.1.8     ✔ stringr 1.5.0
✔ tidyr   1.3.0     ✔ forcats 1.0.0
✔ readr   2.1.4     

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::

In [ ]:
df_flights <- flights
options(repr.matrix.max.cols=50, repr.matrix.max.rows=10)
df_flights

year month day dep_time sched_dep_time dep_delay arr_time sched_arr_time
1      2013 1     1   517      515             2         830      819          
2      2013 1     1   533      529             4         850      830          
3      2013 1     1   542      540             2         923      850          
4      2013 1     1   544      545            -1        1004     1022          
5      2013 1     1   554      600            -6         812      837          
⋮      ⋮    ⋮     ⋮   ⋮        ⋮              ⋮         ⋮        ⋮             
336772 2013 9     30  NA       1455           NA        NA       1634          
336773 2013 9     30  NA       2200           NA        NA       2312          
336774 2013 9     30  NA       1210           NA        NA       1330          
336775 2013 9     30  NA       1159           NA        NA       1344          
336776 2013 9     30  NA        840           NA        NA       1020          
       arr_delay carrier flight tailnum origin dest air_time distance hour
1       11       UA      1545   N14228  EWR    IAH  227      1400     5   
2       20       UA      1714   N24211  LGA    IAH  227      1416     5   
3       33       AA      1141   N619AA  JFK    MIA  160      1089     5   
4      -18       B6       725   N804JB  JFK    BQN  183      1576     5   
5      -25       DL       461   N668DN  LGA    ATL  116       762     6   
⋮      ⋮         ⋮       ⋮      ⋮       ⋮      ⋮    ⋮        ⋮        ⋮   
336772 NA        9E      3393   NA      JFK    DCA  NA       213      14  
336773 NA        9E      3525   NA      LGA    SYR  NA       198      22  
336774 NA        MQ      3461   N535MQ  LGA    BNA  NA       764      12  
336775 NA        MQ      3572   N511MQ  LGA    CLE  NA       419      11  
336776 NA        MQ      3531   N839MQ  LGA    RDU  NA       431       8  
       minute time_hour          
1      15     2013-01-01 05:00:00
2      29     2013-01-01 05:00:00
3      40     2013-01-01 05:00:00
4      45     2013-01-01 05:00:00
5       0     2013-01-01 06:00:00
⋮      ⋮      ⋮                  
336772 55     2013-09-30 14:00:00
336773  0     2013-09-30 22:00:00
336774 10     2013-09-30 12:00:00
336775 59     2013-09-30 11:00:00
336776 40     2013-09-30 08:00:00

In [ ]:
df_airlines <- airlines
df_airlines

carrier name                  
1  9E      Endeavor Air Inc.     
2  AA      American Airlines Inc.
3  AS      Alaska Airlines Inc.  
4  B6      JetBlue Airways       
5  DL      Delta Air Lines Inc.  
⋮  ⋮       ⋮                     
12 UA      United Air Lines Inc. 
13 US      US Airways Inc.       
14 VX      Virgin America        
15 WN      Southwest Airlines Co.
16 YV      Mesa Airlines Inc.

In [ ]:
df_airports <- airports
df_airports

faa name                                 lat      lon        alt  tz dst
1    04G Lansdowne Airport                    41.13047  -80.61958 1044 -5 A  
2    06A Moton Field Municipal Airport        32.46057  -85.68003  264 -6 A  
3    06C Schaumburg Regional                  41.98934  -88.10124  801 -6 A  
4    06N Randall Airport                      41.43191  -74.39156  523 -5 A  
5    09J Jekyll Island Airport                31.07447  -81.42778   11 -5 A  
6    0A9 Elizabethton Municipal Airport       36.37122  -82.17342 1593 -5 A  
7    0G6 Williams County Airport              41.46731  -84.50678  730 -5 A  
8    0G7 Finger Lakes Regional Airport        42.88356  -76.78123  492 -5 A  
9    0P2 Shoestring Aviation Airfield         39.79482  -76.64719 1000 -5 U  
10   0S9 Jefferson County Intl                48.05381 -122.81064  108 -8 A  
11   0W3 Harford County Airport               39.56684  -76.20240  409 -5 A  
12   10C Galt Field Airport                   42.40289  -88.37511  875 -6 U  
13   17G Port Bucyrus-Crawford County Airport 40.78156  -82.97481 1003 -5 A  
14   19A Jackson County Airport               34.17586  -83.56160  951 -5 U  
15   1A3 Martin Campbell Field Airport        35.01581  -84.34683 1789 -5 A  
16   1B9 Mansfield Municipal                  42.00013  -71.19677  122 -5 A  
17   1C9 Frazier Lake Airpark                 54.01333 -124.76833  152 -8 A  
18   1CS Clow International Airport           41.69597  -88.12923  670 -6 U  
19   1G3 Kent State Airport                   41.15139  -81.41511 1134 -5 A  
20   1G4 Grand Canyon West Airport            35.89990 -113.81567 4813 -7 A  
21   1H2 Effingham Memorial Airport           39.07000  -88.53400  585 -6 A  
22   1OH Fortman Airport                      40.55533  -84.38662  885 -5 U  
23   1RL Point Roberts Airpark                48.97972 -123.07889   10 -8 A  
24   23M Clarke CO                            32.05170  -88.44340  320 -6 A  
25   24C Lowell City Airport                  42.95392  -85.34391  681 -5 A  
26   24J Suwannee County Airport              30.30013  -83.02469  104 -5 A  
27   25D Forest Lake Airport                  45.24775  -92.99439  925 -6 A  
28   29D Grove City Airport                   41.14603  -80.16775 1371 -5 A  
29   2A0 Mark Anton Airport                   35.48625  -84.93108  718 -5 A  
30   2B2 Plum Island Airport                  42.79536  -70.83944   11 -5 A  
⋮    ⋮   ⋮                                    ⋮        ⋮          ⋮    ⋮  ⋮  
1429 X59 Valkaria Municipal                   27.96086  -80.55833   26 -5 A  
1430 XFL Flagler County Airport               29.28210  -81.12120   33 -5 A  
1431 XNA NW Arkansas Regional                 36.28187  -94.30681 1287 -6 A  
1432 XZK Amherst Amtrak Station AMM           42.37500  -72.51139  258 -5 A  
1433 Y51 Municipal Airport                    43.57936  -90.89647 1292 -6 A  
1434 Y72 Bloyer Field                         43.97622  -90.48061  966 -6 A  
1435 YAK Yakutat                              59.30120 -139.39370   33 -9 A  
1436 YIP Willow Run                           42.23793  -83.53041  716 -5 A  
1437 YKM Yakima Air Terminal McAllister Field 46.56820 -120.54400 1095 -8 A  
1438 YKN Chan Gurney                          42.87110  -97.39690 1200 -6 A  
1439 YNG Youngstown Warren Rgnl               41.26074  -80.67910 1196 -5 A  
1440 YUM Yuma Mcas Yuma Intl                  32.65658 -114.60598  216 -7 N  
1441 Z84 Clear                                64.30120 -149.12014  552 -9 A  
1442 ZBP Penn Station                         39.30722  -76.61556   66 -5 A  
1443 ZFV Philadelphia 30th St Station         39.95570  -75.18200    0 -5 A  
1444 ZPH Municipal Airport                    28.22806  -82.15583   90 -5 A  
1445 ZRA Atlantic City Rail Terminal          39.36650  -74.44200    8 -5 A  
1446 ZRD Train Station                        37.53430  -77.42945   26 -5 A  
1447 ZRP Newark Penn Station                  40.73472  -74.16417    0 -5 A  
1448 ZRT Hartford Union Sta

In [ ]:
df_planes <- planes
df_planes 

tailnum year type                    manufacturer                 
1    N10156  2004 Fixed wing multi engine EMBRAER                      
2    N102UW  1998 Fixed wing multi engine AIRBUS INDUSTRIE             
3    N103US  1999 Fixed wing multi engine AIRBUS INDUSTRIE             
4    N104UW  1999 Fixed wing multi engine AIRBUS INDUSTRIE             
5    N10575  2002 Fixed wing multi engine EMBRAER                      
6    N105UW  1999 Fixed wing multi engine AIRBUS INDUSTRIE             
7    N107US  1999 Fixed wing multi engine AIRBUS INDUSTRIE             
8    N108UW  1999 Fixed wing multi engine AIRBUS INDUSTRIE             
9    N109UW  1999 Fixed wing multi engine AIRBUS INDUSTRIE             
10   N110UW  1999 Fixed wing multi engine AIRBUS INDUSTRIE             
11   N11106  2002 Fixed wing multi engine EMBRAER                      
12   N11107  2002 Fixed wing multi engine EMBRAER                      
13   N11109  2002 Fixed wing multi engine EMBRAER                      
14   N11113  2002 Fixed wing multi engine EMBRAER                      
15   N11119  2002 Fixed wing multi engine EMBRAER                      
16   N11121  2003 Fixed wing multi engine EMBRAER                      
17   N11127  2003 Fixed wing multi engine EMBRAER                      
18   N11137  2003 Fixed wing multi engine EMBRAER                      
19   N11140  2003 Fixed wing multi engine EMBRAER                      
20   N11150  2003 Fixed wing multi engine EMBRAER                      
21   N11155  2004 Fixed wing multi engine EMBRAER                      
22   N11164  2004 Fixed wing multi engine EMBRAER                      
23   N11165  2004 Fixed wing multi engine EMBRAER                      
24   N11176  2004 Fixed wing multi engine EMBRAER                      
25   N11181  2005 Fixed wing multi engine EMBRAER                      
26   N11184  2005 Fixed wing multi engine EMBRAER                      
27   N11187  2005 Fixed wing multi engine EMBRAER                      
28   N11189  2005 Fixed wing multi engine EMBRAER                      
29   N11191  2005 Fixed wing multi engine EMBRAER                      
30   N11192  2005 Fixed wing multi engine EMBRAER                      
⋮    ⋮       ⋮    ⋮                       ⋮                            
3293 N984DL  1991 Fixed wing multi engine MCDONNELL DOUGLAS AIRCRAFT CO
3294 N985AT  2001 Fixed wing multi engine BOEING                       
3295 N985DL  1991 Fixed wing multi engine MCDONNELL DOUGLAS AIRCRAFT CO
3296 N986AT  2001 Fixed wing multi engine BOEING                       
3297 N986DL  1991 Fixed wing multi engine MCDONNELL DOUGLAS AIRCRAFT CO
3298 N987AT  2001 Fixed wing multi engine BOEING                       
3299 N987DL  1991 Fixed wing multi engine MCDONNELL DOUGLAS AIRCRAFT CO
3300 N988AT  2001 Fixed wing multi engine BOEING                       
3301 N988DL  1991 Fixed wing multi engine MCDONNELL DOUGLAS AIRCRAFT CO
3302 N989AT  2001 Fixed wing multi engine BOEING                       
3303 N989DL  1991 Fixed wing multi engine MCDONNELL DOUGLAS AIRCRAFT CO
3304 N990AT  2001 Fixed wing multi engine BOEING                       
3305 N990DL  1991 Fixed wing multi engine MCDONNELL DOUGLAS AIRCRAFT CO
3306 N991AT    NA Fixed wing multi engine BOEING                       
3307 N991DL  1991 Fixed wing multi engine MCDONNELL DOUGLAS AIRCRAFT CO
3308 N992AT  2002 Fixed wing multi engine BOEING                       
3309 N992DL  1991 Fixed wing multi engine MCDONNELL DOUGLAS AIRCRAFT CO
3310 N993AT  2002 Fixed wing multi engine BOEING                       
3311 N993DL  1991 Fixed wing multi engine MCDONNELL DOUGLAS AIRCRAFT CO
3312 N994AT  2002 Fixed wing multi engine BOEING                       
3313 N994DL  1991 Fixed wing multi engine MCDONNELL DOUGLAS CORPORATION
3314 N995AT  2002 Fixed wing multi engine BOEING                       
3315 N995DL  1991 Fixed wing multi engine MCDONNELL DOUGLAS AIRCRAFT CO
3316 N996AT  2002 Fixed wing multi engine BOE

In [ ]:
df_weather <- weather
df_weather

origin year month day hour temp  dewp  humid wind_dir wind_speed
1     EWR    2013 1     1    1   39.02 26.06 59.37 270      10.35702  
2     EWR    2013 1     1    2   39.02 26.96 61.63 250       8.05546  
3     EWR    2013 1     1    3   39.02 28.04 64.43 240      11.50780  
4     EWR    2013 1     1    4   39.92 28.04 62.21 250      12.65858  
5     EWR    2013 1     1    5   39.02 28.04 64.43 260      12.65858  
6     EWR    2013 1     1    6   37.94 28.04 67.21 240      11.50780  
7     EWR    2013 1     1    7   39.02 28.04 64.43 240      14.96014  
8     EWR    2013 1     1    8   39.92 28.04 62.21 250      10.35702  
9     EWR    2013 1     1    9   39.92 28.04 62.21 260      14.96014  
10    EWR    2013 1     1   10   41.00 28.04 59.65 260      13.80936  
11    EWR    2013 1     1   11   41.00 26.96 57.06 260      14.96014  
12    EWR    2013 1     1   13   39.20 28.40 69.67 330      16.11092  
13    EWR    2013 1     1   14   39.02 24.08 54.68 280      13.80936  
14    EWR    2013 1     1   15   37.94 24.08 57.04 290       9.20624  
15    EWR    2013 1     1   16   37.04 19.94 49.62 300      13.80936  
16    EWR    2013 1     1   17   35.96 19.04 49.83 330      11.50780  
17    EWR    2013 1     1   18   33.98 15.08 45.43 310      12.65858  
18    EWR    2013 1     1   19   33.08 12.92 42.84 320      10.35702  
19    EWR    2013 1     1   20   32.00 15.08 49.19 310      14.96014  
20    EWR    2013 1     1   21   30.02 12.92 48.48 320      18.41248  
21    EWR    2013 1     1   22   28.94 12.02 48.69 320      18.41248  
22    EWR    2013 1     1   23   28.04 10.94 48.15 310      16.11092  
23    EWR    2013 1     2    0   26.96 10.94 50.34 310      14.96014  
24    EWR    2013 1     2    1   26.06 10.94 52.25 330      12.65858  
25    EWR    2013 1     2    2   24.98 10.94 54.65 330      13.80936  
26    EWR    2013 1     2    3   24.08  8.96 51.93 320      14.96014  
27    EWR    2013 1     2    4   24.08  8.96 51.93 330      12.65858  
28    EWR    2013 1     2    5   24.08  8.96 51.93 330       6.90468  
29    EWR    2013 1     2    6   24.08  8.96 51.93 310       3.45234  
30    EWR    2013 1     2    7   24.98 10.04 52.50 300       6.90468  
⋮     ⋮      ⋮    ⋮     ⋮   ⋮    ⋮     ⋮     ⋮     ⋮        ⋮         
26086 LGA    2013 12    29  13   42.80 37.94 88.76  70      12.65858  
26087 LGA    2013 12    29  14   41.00 37.94 93.19  60      18.41248  
26088 LGA    2013 12    29  15   41.00 39.02 92.59  40      13.80936  
26089 LGA    2013 12    29  16   41.00 37.94 88.76 350       8.05546  
26090 LGA    2013 12    29  17   44.06 41.00 93.24 350      20.71404  
26091 LGA    2013 12    29  18   42.08 39.02 88.81 330      14.96014  
26092 LGA    2013 12    29  19   42.80 37.94 85.13 320      17.26170  
26093 LGA    2013 12    29  20   42.08 37.94 86.89 320      19.56326  
26094 LGA    2013 12    29  21   42.80 37.40 82.17 320      16.11092  
26095 LGA    2013 12    29  22   42.98 37.04 79.38 300       9.20624  
26096 LGA    2013 12    29  23   42.98 35.06 73.39 310      17.26170  
26097 LGA    2013 12    30   0   42.08 33.98 72.78 320      11.50780  
26098 LGA    2013 12    30   1   42.08 33.98 72.78 250       9.20624  
26099 LGA    2013 12    30   2   41.00 33.98 75.88 240       8.05546  
26100 LGA    2013 12    30   3   42.98 33.98 70.30 270       9.20624  
26101 LGA    2013 12    30   4   41.00 33.08 73.19   0       0.00000  
26102 LGA    2013 12    30   5   42.98 33.08 67.81 250      10.35702  
26103 LGA    2013 12    30   6   42.98 33.98 70.30 230       6.90468  
26104 LGA    2013 12    30   7   44.06 35.06 70.42 240      11.50780  
26105 LGA    2013 12    30   8   44.06 33.98 67.45 260      11.50780  
26106 LGA    2013 12    30   9   44.06 33.08 65.07 260      13.80936  
26107 LGA    2013 12    30  10   42.98 33.80 70.28 330      16.11092  
26108 LGA    2013 12    30  11   41.00 28.40 62.21 340      13.80936  
26109 LGA    2013 12    30  12   37.94 23.00 54.51 330      21.86482  
26110 LGA    2013 12    30  13

In [ ]:
df_f <- flights %>% 
  select(origin,dest,distance) 
df_f

origin dest distance
1      EWR    IAH  1400    
2      LGA    IAH  1416    
3      JFK    MIA  1089    
4      JFK    BQN  1576    
5      LGA    ATL   762    
⋮      ⋮      ⋮    ⋮       
336772 JFK    DCA  213     
336773 LGA    SYR  198     
336774 LGA    BNA  764     
336775 LGA    CLE  419     
336776 LGA    RDU  431

What is the average distance for each months ? lets find the cheapest month to travel

In [ ]:
mean_distance <- flights %>%
    select (month,origin,dest,distance) %>%
    group_by(month) %>%
    summarize(average_distance = mean(distance)) %>%
    arrange(desc(average_distance)) 
mean_distance
head(5)

month average_distance
1  12    1064.656        
2   8    1062.138        
3   7    1058.596        
4   6    1057.125        
5  11    1050.305        
⋮  ⋮     ⋮               
8  10    1038.876        
9   4    1038.733        
10  3    1011.987        
11  1    1006.844        
12  2    1000.982

In [ ]:
mean_distance <- flights %>%
    select (month,origin,dest,distance) %>%
    group_by(month) %>%
    summarize(average_distance = mean(distance)) %>%
    arrange(desc(average_distance)) %>%
    head(5)
mean_distance


month average_distance
1 12    1064.656        
2  8    1062.138        
3  7    1058.596        
4  6    1057.125        
5 11    1050.305

most people => 
  month #12 >> travel time ,
  month #1,2 >> work time 

---




Since December is the most popular month to travel , where do they go?

In [ ]:
top_month <- flights %>% 
  filter(month == 12) %>%
  # select(dest) %>%
  count(dest) %>%
  arrange(desc(n))

top_month 

dest n   
1  ATL  1463
2  LAX  1408
3  MCO  1219
4  ORD  1196
5  CLT  1193
⋮  ⋮    ⋮   
92 BZN  4   
93 HDN  2   
94 MTJ  2   
95 PSP  2   
96 SBN  1

The most common attraction is ATL aka atlanta airport, hmm... I wonder what are they going there for? Personally i thought LAX (clout city) would be the most popular,however, i am wrong 

---



What airline has the highest air time ?

In [ ]:
highest_airtime  <- flights %>%
  
  filter(!is.na(air_time)) %>%
  arrange(desc(air_time))  %>% 
  left_join(airlines, by = "carrier") %>%
  select(airline = name,
         ID = carrier,
         air_time  ) %>%
  head(8)
highest_airtime

airline                ID air_time
1 United Air Lines Inc.  UA 695     
2 Hawaiian Airlines Inc. HA 691     
3 Hawaiian Airlines Inc. HA 686     
4 Hawaiian Airlines Inc. HA 686     
5 Hawaiian Airlines Inc. HA 683     
6 Hawaiian Airlines Inc. HA 679     
7 United Air Lines Inc.  UA 676     
8 Hawaiian Airlines Inc. HA 676

United Air Lines Inc.	(UA) won first place and recieved the prize while majority of the airline placed in top 8 is Hawaiian Airlines Inc. (HA)

---



what airline is the Worst?

In [ ]:
best_airline <- flights %>% 
  left_join(airlines, by = "carrier") %>%
  select(name, carrier, dep_delay) %>%
  filter(!is.na(dep_delay)) %>%
  group_by(name, carrier) %>%
  summarise(sum_dep_delay = sum(dep_delay), filght_count = n()) %>%
  mutate(delay_per_flight = sum_dep_delay / filght_count) %>%
  arrange(desc(delay_per_flight))%>%
  head(8)

best_airline

`summarise()` has grouped output by 'name'. You can override using the
`.groups` argument.


name                        carrier sum_dep_delay filght_count
1 Frontier Airlines Inc.      F9        13787         682       
2 ExpressJet Airlines Inc.    EV      1024829       51356       
3 Mesa Airlines Inc.          YV        10353         545       
4 AirTran Airways Corporation FL        59680        3187       
5 Southwest Airlines Co.      WN       214011       12083       
6 Endeavor Air Inc.           9E       291296       17416       
7 JetBlue Airways             B6       705417       54169       
8 Virgin America              VX        66033        5131       
  delay_per_flight
1 20.21554        
2 19.95539        
3 18.99633        
4 18.72607        
5 17.71174        
6 16.72577        
7 13.02252        
8 12.86942

the absolute worst airline is Frontier Airlines Inc.	which average 20.21554 for (total) delay_per flight, please choose anything but this airline TwT

---



Lets make a slight detour toward the weather table/data, can't wait for winter season in 2023 , lets see why winter season is the best 

In [ ]:
why_winter_best <- weather %>%
    select (month, temp) %>%
    filter(!is.na(temp)) %>%
    group_by(month) %>%
    summarize(mean_temp = mean(temp), sd_temp = sd(temp)) %>%
    dplyr::arrange(mean_temp)%>%
    head(3)
why_winter_best

month mean_temp sd_temp  
1  2    34.27060   6.982378
2  1    35.63566  10.224635
3 12    38.44180   9.982432

that goes without saying, 3 best months of the year ,winter season, surely is the best. Thats probably why most people travels in this period too
